grant-project/grant-monorepo
# Working Refunds

Closes #10.

### What This Does

Hooks up the refunding view under the governance tab. Allows users to see an up-to-date count of refund votes, make a vote if they've contributed, and receive their refund if the vote has passed.

### What Needs Work

Right now the contract is only refundable if it's "frozen," which only happens if someone runs the `refund` function after it's entered a state where it could be. This doesn't happen automatically, and I didn't build out a UI for allowing anyone to run this arbitrarily. Ideally the backend would automatically run this for everyone, rather than have a user do it.

Likewise, there are states where you can get a refund that aren't from a vote. A trustee could force a refund, or the deadline might pass. However, the refund UI only shows up in the case of a voted refund.

In general, all these states should auto-refund from the backend, rather than have users individually call it. Would like input (cc @dternyak) for how this PR should proceed.

### Steps to Test

* Start a proposal with account 1
* Have account 2 fund it by 40%, account 3 by 60%.
* Have account 2 vote for a refund. Confirm it shows up, but that you can't request a refund.
* Have account 3 vote for a refund. Confirm it shows up, and that both accounts can now request a refund.
* Have accounts 2 and 3 request a refund. Confirm it works.
* Confirm that a second refund cannot be requested.
